src/relations/mysql.py

```diff
@@ -0,0 +1,146 @@
+    def __init__(self, charm):
+        super().__init__(charm, LEGACY_MYSQL)
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(
+            self.charm.on[LEGACY_MYSQL].relation_created, self._on_mysql_relation_created
+        )
+
+        self.framework.observe(
+            self.charm.on[LEGACY_MYSQL].relation_broken, self._on_mysql_relation_broken
+        )
+
+    def _get_or_set_password_in_peer_databag(self, username: str) -> str:
+        """Get a user's password from the peer databag if it exists, else populate a password.
+
+        Args:
+            username: The mysql username
+
+        Returns:
+            a string representing the password for the mysql user
+        """
+        peer_databag = self.charm._peers.data[self.charm.app]
+
+        if peer_databag.get(f"{username}_password"):
+            return peer_databag.get(f"{username}_password")
+
+        password = generate_random_password(PASSWORD_LENGTH)
+        peer_databag[f"{username}_password"] = password
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
+        """
+        # Skip if the charm is not past the setup phase (config-changed event not executed yet)
+        if not self.charm._is_peer_data_set:
+            return
+
+        relation_data = json.loads(
+            self.charm._peers.data[self.charm.app].get("mysql_relation_data", "{}")
+        )
+
+        for relation in self.charm.model.relations.get(LEGACY_MYSQL, []):
+            relation_databag = relation.data
+
+            # Copy relation data into the new leader unit's databag
+            for key, value in relation_data.items():
+                if relation_databag[self.charm.unit].get(key) != value:
+                    relation_databag[self.charm.unit][key] = value
+
+            # Assign the cluster primary's address as the database host
+            primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+            relation_databag[self.charm.unit]["host"] = primary_address
+
+    def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the legacy 'mysql' relation created event.
+
+        Will set up the database and the scoped application user. The connection
+        data (relation data) is then copied into the peer relation databag (to
+        be copied over to the new leader unit's databag in case of a new leader
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
```

Comment:
it is possible that this event will defer continuously and block the `on-config-changed` event from executing, or not block it but lead to a long time for that hook to actually get fired. This will lead to poor performance. Perhaps it would be best to:

```
    def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
        ....
        
        if not self.charm._is_peer_data_set:
            self.app_data['needs relation created']=True
```
and then 
```
     def config_changed -> None:
        ....
        
        if self.app_data['needs relation created']:
            <emit event or call this function>
           self.app_data['needs relation created'] = False
```

Reply:
this kind of delay will only happen on initial run, if the apps are related while still on first deploying (e.g. when packed in a bundle). A single update-status interval will be enough.